src/ai/backend/manager/server.py

```diff
@@ -490,6 +491,113 @@ async def monitoring_ctx(root_ctx: RootContext) -> AsyncIterator[None]:
+            .where(SessionRow.status == status)
+            .where(
+                (
+                    datetime.now(tz=tzutc())
+                    - SessionRow.status_history[status.name].astext.cast(
+                        sa.types.DateTime(timezone=True)
+                    )
+                )
+                > threshold
+            )
+            .options(
+                noload("*"),
+                load_only(SessionRow.id, SessionRow.name, SessionRow.status, SessionRow.access_key),
+            )
+        )
+        async with db.begin_readonly() as conn:
+            result = await conn.execute(query)
+            return result.fetchall()
+
+    async def _force_terminate_hanging_sessions(
+        status: SessionStatus,
+        threshold: timedelta,
+        reason: KernelLifecycleEventReason = KernelLifecycleEventReason.HANGTIME_EXCEEDED,
+    ) -> None:
+        while True:
+            sessions = await _fetch_hanging_sessions(
+                root_ctx.db, status=status, threshold=threshold
+            )
+            log.debug(f"{len(sessions)} {status} sessions found.")
+
+            _ = await asyncio.gather(
+                *[
+                    asyncio.create_task(
+                        root_ctx.registry.destroy_session(session, forced=True, reason=reason),
+                    )
+                    for session in sessions
+                ]
+            )
+
+            await asyncio.sleep(threshold.seconds)
+
+    raw_session_config = await root_ctx.shared_config.etcd.get_prefix("config/session")
+    assert isinstance(raw_session_config, Mapping)
+
+    hang_tolerance_threshold_dict = raw_session_config.get("hang-tolerance-threshold", {})
+    assert isinstance(hang_tolerance_threshold_dict, Mapping)
+
+    session_force_termination_tasks = []
+    for status, threshold_fmt in hang_tolerance_threshold_dict.items():
+        assert isinstance(threshold_fmt, str)
+        try:
+            session_status = SessionStatus[status]
+        except KeyError:
+            continue
+
+        match threshold_fmt[-1]:
+            case "h":
+                threshold = timedelta(hours=int(threshold_fmt[:-1]))
+            case "m":
+                threshold = timedelta(minutes=int(threshold_fmt[:-1]))
+            case "s":
+                threshold = timedelta(seconds=int(threshold_fmt[:-1]))
+            case _:
+                log.error(
+                    f'Skip "{status}:{threshold_fmt}" as it is an invalid argument. Supported'
+                    ' postfixes: "h", "m", "s"'
+                )
+                continue
+
+        session_force_termination_tasks.append(
+            asyncio.create_task(
+                _force_terminate_hanging_sessions(status=session_status, threshold=threshold)
+            )
+        )
+
+    yield
+
+    for task in session_force_termination_tasks:
+        if not task.done():
+            if not task.cancelled():
+                task.cancel()
```

Comment:
We may just use `if not task.done()` only here.

Reply:
Resolved: f2ebf3d4e8277618731f8c01e967ef1de283d087